### PR TITLE
Support schema for List, Set, Map and Array of objects

### PR DESCRIPTION
**Pull request by gs-snagaraj**
## Issue

Closes ##1480 [BUG] A list of objects is not serialised correctly
## Change
DefaultToolExecutor now handles List, Array, Set, Map 

## General checklist

- [X] There are no breaking changes
- [X] I have added unit and integration tests for my change
- [X] I have manually run all the unit and integration tests in the module I have added/changed, and they are all green
- [ ] I have manually run all the unit and integration tests in the [core](https://github.com/langchain4j/langchain4j/tree/main/langchain4j-core) and [main](https://github.com/langchain4j/langchain4j/tree/main/langchain4j) modules, and they are all green

- [ ] I have added/updated the [documentation](https://github.com/langchain4j/langchain4j/tree/main/docs/docs)
- [ ] I have added an example in the [examples repo](https://github.com/langchain4j/langchain4j-examples) (only for "big" features)
- [ ] I have added/updated [Spring Boot starter(s)](https://github.com/langchain4j/langchain4j-spring) (if applicable)

## Checklist for adding new model integration

- [ ] I have added my new module in the [BOM](https://github.com/langchain4j/langchain4j/blob/main/langchain4j-bom/pom.xml)

## Checklist for adding new embedding store integration

- [ ] I have added a `{NameOfIntegration}EmbeddingStoreIT` that extends from either `EmbeddingStoreIT` or `EmbeddingStoreWithFilteringIT`
- [ ] I have added my new module in the [BOM](https://github.com/langchain4j/langchain4j/blob/main/langchain4j-bom/pom.xml)

## Checklist for changing existing embedding store integration

- [ ] I have manually verified that the `{NameOfIntegration}EmbeddingStore` works correctly with the data persisted using the latest released version of LangChain4j

### CODE REVIEW

File: langchain4j/src/test/java/dev/langchain4j/agent/tool/DefaultToolExecutorTest.java [hunk 1]
```diff
@@ -296,4 +294,88 @@ public void should_not_execute_tool_with_null_execution_request() throws NoSuchM
...
+
+        @Tool
+        public Person[] saveArray(Person[] arg) {
+            assert arg != null;
+            assert arg.length == 2;
+            return arg;
+        }
+
+        @Tool
+        public List<Person> saveList(List<Person> personList) {
```
Comment (langchain4j) [inline langchain4j/src/test/java/dev/langchain4j/agent/tool/DefaultToolExecutorTest.java]: This does not work, the list contains maps, not `Person`s

File: langchain4j/src/test/java/dev/langchain4j/agent/tool/DefaultToolExecutorTest.java [hunk 2]
```diff
@@ -296,4 +294,88 @@ public void should_not_execute_tool_with_null_execution_request() throws NoSuchM
...
+            assert personList != null;
+            assert personList.size() == 2;
+            assert personList.get(0) != null;
+            return personList;
+        }
+
+        @Tool
+        public Set<Person> saveSet(Set<Person> personSet) {
```
Comment (langchain4j) [inline langchain4j/src/test/java/dev/langchain4j/agent/tool/DefaultToolExecutorTest.java]: This does not work, the set contains maps, not `Person`s

File: langchain4j/src/test/java/dev/langchain4j/agent/tool/DefaultToolExecutorTest.java [hunk 3]
```diff
@@ -296,4 +294,88 @@ public void should_not_execute_tool_with_null_execution_request() throws NoSuchM
...
+            assert personSet != null;
+            assert personSet.size() == 2;
+            return personSet;
+        }
+
+        @Tool
+        public Map<String,Person> saveMap (Map<String,Person> idPersonMap) {
```
Comment (langchain4j) [inline langchain4j/src/test/java/dev/langchain4j/agent/tool/DefaultToolExecutorTest.java]: This does not work, the map contains maps, not `Person`s

File: langchain4j/src/test/java/dev/langchain4j/agent/tool/DefaultToolExecutorTest.java [hunk 4]
```diff
@@ -296,4 +294,88 @@ public void should_not_execute_tool_with_null_execution_request() throws NoSuchM
...
 
     }
-}
\ No newline at end of file
+
+    private static class PersonTool {
+
+        @Tool
+        public Person save( Person arg) {
+            assert arg != null;
```
Comment (langchain4j) [inline langchain4j/src/test/java/dev/langchain4j/agent/tool/DefaultToolExecutorTest.java]: please add assertions for person fields (in all tool methods)

File: langchain4j/src/main/java/dev/langchain4j/agent/tool/DefaultToolExecutor.java [hunk 5]
```diff
@@ -109,14 +106,81 @@ static Object[] prepareArguments(
...
 
+    @SneakyThrows
+    static Object coerceCollectionArgument(
+            Object argument,
+            String parameterName,
+            Class<?> parameterType,
+            Object parameterizedType
+    )  {
+        String json = Json.toJson(argument);
+        if (parameterType == List.class) {
```
Comment (langchain4j) [inline langchain4j/src/main/java/dev/langchain4j/agent/tool/DefaultToolExecutor.java]: 4 blocks of logic in this method does not seem to be related, perhaprs it would be better to have 4 methods?

File: langchain4j/src/main/java/dev/langchain4j/agent/tool/DefaultToolExecutor.java [hunk 6]
```diff
@@ -109,14 +106,81 @@ static Object[] prepareArguments(
...
+            List objects = Json.fromJson(json , List.class);
+            Class<?> elementType  = Class.forName( ((ParameterizedType)parameterizedType).getActualTypeArguments()[0].getTypeName());
```
Comment (langchain4j) [inline langchain4j/src/main/java/dev/langchain4j/agent/tool/DefaultToolExecutor.java]: Is there a way to avoid calling `Class.forName()`?

**Review by ghost — COMMENTED** [review]
@gs-snagaraj thank you!

File: langchain4j/src/test/java/dev/langchain4j/agent/tool/DefaultToolExecutorTest.java (hunk 1 above)
Comment (gs-snagaraj) [reply langchain4j/src/test/java/dev/langchain4j/agent/tool/DefaultToolExecutorTest.java]: My bad, Missed an `else`. Fixed now.

File: langchain4j/src/main/java/dev/langchain4j/agent/tool/DefaultToolExecutor.java (hunk 6 above)
Comment (gs-snagaraj) [reply langchain4j/src/main/java/dev/langchain4j/agent/tool/DefaultToolExecutor.java]: Nope. But it should not fail. Because, It would have failed much before, if the class was not present.

File: langchain4j/src/main/java/dev/langchain4j/agent/tool/DefaultToolExecutor.java [hunk 7]
```diff
@@ -109,14 +106,81 @@ static Object[] prepareArguments(
             if (argumentsMap.containsKey(parameterName)) {
                 Object argument = argumentsMap.get(parameterName);
                 Class<?> parameterType = parameters[i].getType();
-
-                arguments[i] = coerceArgument(argument, parameterName, parameterType);
+                if( parameterType == List.class || parameterType == Set.class || parameterType == Map.class)  {
```
Comment (Wadimz) [inline langchain4j/src/main/java/dev/langchain4j/agent/tool/DefaultToolExecutor.java]: Hi @gs-snagaraj!  Can this check be more generic, than List and Set? Ex. checking ```Collection.class.isAssignableFrom(parameterType)``` instead of `List` and `Set` and handling parameterType as a `Collection`?

Comment (gs-snagaraj) [reply langchain4j/src/main/java/dev/langchain4j/agent/tool/DefaultToolExecutor.java]: Done

File: langchain4j-core/src/main/java/dev/langchain4j/internal/Json.java [hunk 8]
```diff
@@ -46,6 +47,16 @@ public interface JsonCodec {
          * @throws IOException if an I/O error occurs.
          */
         InputStream toInputStream(Object o, Class<?> type) throws IOException;
+
+        /**
+         * Convert the given JSON string to an object of the given type.
+         * @param json
+         * @param type
+         * @return the object.
+         * @param <T>
```
Comment (Wadimz) [inline langchain4j-core/src/main/java/dev/langchain4j/internal/Json.java]: Looks code doesn't have javadocs where parameters are present without description. Can we remove them or provide description?
nit: extra space after javadoc

File: langchain4j/src/main/java/dev/langchain4j/agent/tool/DefaultToolExecutor.java [hunk 9]
```diff
@@ -109,14 +108,52 @@ static Object[] prepareArguments(
             if (argumentsMap.containsKey(parameterName)) {
                 Object argument = argumentsMap.get(parameterName);
                 Class<?> parameterType = parameters[i].getType();
-
-                arguments[i] = coerceArgument(argument, parameterName, parameterType);
+                if( Collection.class.isAssignableFrom(parameterType) || Map.class.isAssignableFrom(parameterType))  {
```
Comment (Wadimz) [inline langchain4j/src/main/java/dev/langchain4j/agent/tool/DefaultToolExecutor.java]: nit: indentations after `if` are different in the same file. Maybe we can align them to be the same to follow similar codestyle?

File: langchain4j/src/test/java/dev/langchain4j/agent/tool/DefaultToolExecutorTest.java [hunk 10]
```diff
@@ -296,4 +294,109 @@ public void should_not_execute_tool_with_null_execution_request() throws NoSuchM
...
+            return arg;
+        }
+
+        @Tool
+        public Person[] saveArray(Person[] arg) {
+            assert arg != null;
```
Comment (Wadimz) [inline langchain4j/src/test/java/dev/langchain4j/agent/tool/DefaultToolExecutorTest.java]: doesn't look `assert` is used somewhere in project's tests. Can we consider using  Assertj library to be consistent?

**Review by Wadimz — COMMENTED** [review]
Thanks for addressing comments, looks great. Added a few minor cosmetic notes

File: langchain4j/src/test/java/dev/langchain4j/agent/tool/DefaultToolExecutorTest.java (hunk 10 above)
Comment (gs-snagaraj) [reply langchain4j/src/test/java/dev/langchain4j/agent/tool/DefaultToolExecutorTest.java]: assert here is to fail the tool execution with error. Not a Junit.

Junit Asserts adding in Test methods.

File: langchain4j/src/main/java/dev/langchain4j/agent/tool/DefaultToolExecutor.java (hunk 9 above)
Comment (gs-snagaraj) [reply langchain4j/src/main/java/dev/langchain4j/agent/tool/DefaultToolExecutor.java]: Did get it. It looks good only.

File: langchain4j-core/src/main/java/dev/langchain4j/internal/Json.java (hunk 8 above)
Comment (gs-snagaraj) [reply langchain4j-core/src/main/java/dev/langchain4j/internal/Json.java]: Done

**Comment by gs-snagaraj** [comment]
@Wadimz @langchain4j  Please check and merge.

**Comment by langchain4j** [comment]
@gs-snagaraj there are conflicts, could you please check?

**Comment by gs-snagaraj** [comment]
@Wadimz @langchain4j  Please check and merge.

File: langchain4j-core/src/main/java/dev/langchain4j/internal/Json.java [hunk 11]
```diff
@@ -90,4 +100,15 @@ public static <T> T fromJson(String json, Class<T> type) {
...
     }
+
+    /**
+     * Convert the given JSON string to an object of the given type.
+     * @param json the JSON string.
+     * @param type the type of the object represented by {@link TypeToken}.
+     * @return the object.
+     * @param <T> the type of the object.
+     */
+    public static <T> T fromJson(String json, TypeToken<T> type) {
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/internal/Json.java:111]: `TypeToken` is a Gson-specific class, but this interface is supposed to be implementation-agnostic. We are also slowly migrating to Jackson, so ideally the solution should work for Jackson as well.
Is it possible to make it work without introducing `TypeToken` into this interface?

**Review by ghost — COMMENTED** [review]
@gs-snagaraj thank you a lot!

File: langchain4j-core/src/main/java/dev/langchain4j/internal/Json.java (hunk 11 above)
Comment (langchain4j) [reply langchain4j-core/src/main/java/dev/langchain4j/internal/Json.java:111]: image
`argument` here is a `List<LinkedTreeMap>`. I guess each `LinkedTreeMap` can (separately) be converted into a `Person` using existing  `Json.fromJson(String json, Class<T> type)` method

Comment (gs-snagaraj) [reply langchain4j-core/src/main/java/dev/langchain4j/internal/Json.java:111]: Tried it but no luck. For Jackson we need to use TypeReference

Comment (langchain4j) [reply langchain4j-core/src/main/java/dev/langchain4j/internal/Json.java:111]: @gs-snagaraj `TypeReference` is needed only when we try to deserialize `List<T>`, right? To deserialize each `T` separately we do not need it.
AFAIU, since `argument` is a `List<T>` where each `T` is `LinkedTreeMap`, we can serialize `LinkedTreeMap` into `String` and deserialize it into T without the need of `TypeReference` or `TypeToken`.

**Comment by langchain4j** [comment]
@gs-snagaraj since I did not get any answer and I am reworking this part of the code, I will close this PR.
It will be fixed as a part of https://github.com/langchain4j/langchain4j/pull/1668.

Thank you for your PR! BTW, I have copied your tests, it helped a lot!
